File: src/claw_mem/values/value_backup.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Value Backup - Values local storage
"""

import json
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Dict, List, Optional

from claw_mem.values import UserValueStore


@dataclass
class BackupMetadata:
    """Backup metadata"""

    user_id: str
    backup_id: str
    created_at: datetime
    file_path: str
    file_size: int
    values_count: int
    checksum: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "user_id": self.user_id,
            "backup_id": self.backup_id,
            "created_at": self.created_at.isoformat(),
            "file_path": self.file_path,
            "file_size": self.file_size,
            "values_count": self.values_count,
            "checksum": self.checksum,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "BackupMetadata":
        return cls(
            user_id=data["user_id"],
            backup_id=data["backup_id"],
            created_at=datetime.fromisoformat(data["created_at"]),
            file_path=data["file_path"],
            file_size=data["file_size"],
            values_count=data["values_count"],
            checksum=data.get("checksum", ""),
        )


class ValueBackup:
    """Values backup manager"""

    def __init__(
        self, value_store: Optional[UserValueStore] = None, backup_dir: Optional[Path] = None
    ):
        """Initialize backup manager

        Args:
            value_store: User values storage
            backup_dir: Backup directory, default ~/.claw_mem/backups/
        """
        self.value_store = value_store or UserValueStore()

        if backup_dir is None:
            backup_dir = Path.home() / ".claw_mem" / "backups"

        self.backup_dir = backup_dir
        self.backup_dir.mkdir(parents=True, exist_ok=True)

        # Metadata file
        self.metadata_file = self.backup_dir / "metadata.json"

    def export_values(self, user_id: str, path: Optional[Path] = None) -> BackupMetadata:
        """Export user values to file

        Args:
            user_id: User ID
            path: Export path, auto-generated if None

        Returns:
            BackupMetadata: Backup metadata
        """
        # Get user values
        user_values = self.value_store.get_user_values(user_id)
        if not user_values:
            raise ValueError(f"User {user_id} not found")

        # Generate backup ID and path
        import uuid

        backup_id = str(uuid.uuid4())[:8]

        if path is None:
            timestamp = datetime.now(timezone.utc).strftime("%Y%m%d_%H%M%S")
            path = self.backup_dir / f"{user_id}_{timestamp}.json"

        # Export data
        export_data = {
            "user_id": user_id,
            "exported_at": datetime.now(timezone.utc).isoformat(),
            "version": "2.2.0",
            "values": user_values.to_dict(),
        }

        # Write file
        path.write_text(json.dumps(export_data, indent=2, ensure_ascii=False), encoding="utf-8")

        # Calculate simple checksum
        checksum = str(abs(hash(str(export_data))))[:16]

        # Create metadata
        metadata = BackupMetadata(
            user_id=user_id,
            backup_id=backup_id,
            created_at=datetime.now(timezone.utc),
            file_path=str(path),
            file_size=path.stat().st_size,
            values_count=len(user_values.principles)
            + len(user_values.preferences)
            + len(user_values.red_lines),
            checksum=checksum,
        )

        # Save metadata
        self._save_metadata(metadata)

        return metadata

    def import_values(self, user_id: str, path: Path, overwrite: bool = False) -> bool:
        """Import user values from file

        Args:
            user_id: User ID
            path: Import file path
            overwrite: Whether to overwrite existing data

        Returns:
            bool: Whether import was successful
        """
        # Read file
        if not path.exists():
            raise FileNotFoundError(f"Backup file not found: {path}")

        content = path.read_text(encoding="utf-8")
        data = json.loads(content)

        if "values" not in data:
            raise ValueError("Invalid backup file format")

        imported_values = data["values"]

        # Check user ID match
        if imported_values.get("user_id") != user_id:
            # Allow importing values from different user (creates new user)
            pass

        # Get existing values
        existing = self.value_store.get_user_values(user_id)

        if existing and not overwrite:
            raise ValueError(f"User {user_id} already exists. Use overwrite=True to replace.")

        # Import principles
        for principle in imported_values.get("principles", []):
            self.value_store.save_principle(user_id, principle)

        # Import preferences
        for key, value in imported_values.get("preferences", {}).items():
            self.value_store.save_preference(user_id, key, value)

        # Import red lines
        for line in imported_values.get("red_lines", []):
            self.value_store.save_red_line(user_id, line)

        return True

    def list_backups(self, user_id: Optional[str] = None) -> List[BackupMetadata]:
        """List backup files

        Args:
            user_id: User ID, or None to list all users' backups

        Returns:
            List[BackupMetadata]: List of backup metadata
        """
        metadata_list = []

        if self.metadata_file.exists():
            try:
                all_metadata = json.loads(self.metadata_file.read_text(encoding="utf-8"))
            except json.JSONDecodeError:
                all_metadata = {}
        else:
            all_metadata = {}

        for backup_id, meta_data in all_metadata.items():
            if user_id is None or meta_data.get("user_id") == user_id:
                try:
                    metadata_list.append(BackupMetadata.from_dict(meta_data))
                except (KeyError, ValueError):
                    continue

        # Sort by time
        metadata_list.sort(key=lambda m: m.created_at, reverse=True)

        return metadata_list

    def backup_metadata(self, user_id: str) -> Dict[str, Any]:
        """Get user backup metadata

        Args:
            user_id: User ID

        Returns:
            Dict: Metadata summary
        """
        backups = self.list_backups(user_id)

        if not backups:
            return {"user_id": user_id, "backup_count": 0, "latest_backup": None, "total_size": 0}

        return {
            "user_id": user_id,
            "backup_count": len(backups),
            "latest_backup": backups[0].to_dict() if backups else None,
            "total_size": sum(b.file_size for b in backups),
        }

    def delete_backup(self, backup_id: str) -> bool:
        """Delete backup

        Args:
            backup_id: Backup ID

        Returns:
            bool: Whether deletion was successful
        """
        if self.metadata_file.exists():
            try:
                all_metadata = json.loads(self.metadata_file.read_text(encoding="utf-8"))
            except json.JSONDecodeError:
                return False
        else:
            return False

        if backup_id not in all_metadata:
            return False

        meta_data = all_metadata[backup_id]
        file_path = Path(meta_data["file_path"])

        # Delete file
        if file_path.exists():
            file_path.unlink()

        # Delete metadata
        del all_metadata[backup_id]
        self.metadata_file.write_text(json.dumps(all_metadata, indent=2), encoding="utf-8")

        return True

    def _save_metadata(self, metadata: BackupMetadata) -> None:
        """Save backup metadata"""
        all_metadata = {}

        if self.metadata_file.exists():
            try:
                all_metadata = json.loads(self.metadata_file.read_text(encoding="utf-8"))
            except json.JSONDecodeError:
                pass

        all_metadata[metadata.backup_id] = metadata.to_dict()

        self.metadata_file.write_text(json.dumps(all_metadata, indent=2), encoding="utf-8")


__all__ = [
    "BackupMetadata",
    "ValueBackup",
]

File: src/values/value_backup.ts
// Copyright 2026 Peter Cheng
// Licensed under Apache-2.0

/**
 * Value Backup - Values local storage
 */

import * as fs from "fs";
import * as path from "path";
import * as os from "os";
import * as crypto from "crypto";
import { UserValueStore, userValueToDict } from "./user_value_store";

// ── Backup metadata type ───────────────────────────────────────────────

export interface BackupMetadata {
  userId: string;
  backupId: string;
  createdAt: Date;
  filePath: string;
  fileSize: number;
  valuesCount: number;
  checksum: string;
}

export function backupMetadataToDict(m: BackupMetadata): Record<string, unknown> {
  return {
    user_id: m.userId,
    backup_id: m.backupId,
    created_at: m.createdAt.toISOString(),
    file_path: m.filePath,
    file_size: m.fileSize,
    values_count: m.valuesCount,
    checksum: m.checksum,
  };
}

export function backupMetadataFromDict(data: Record<string, unknown>): BackupMetadata {
  return {
    userId: data.user_id as string,
    backupId: data.backup_id as string,
    createdAt: new Date(data.created_at as string),
    filePath: data.file_path as string,
    fileSize: data.file_size as number,
    valuesCount: data.values_count as number,
    checksum: (data.checksum as string) || "",
  };
}

// ── Backup manager ─────────────────────────────────────────────────────

export class ValueBackup {
  private _metadataFile: string;

  constructor(
    private _valueStore: UserValueStore = new UserValueStore(),
    backupDir?: string,
  ) {
    const dir = backupDir || path.join(os.homedir(), ".claw_mem", "backups");
    fs.mkdirSync(dir, { recursive: true });
    this.backupDir = dir;
    this._metadataFile = path.join(dir, "metadata.json");
  }

  backupDir: string;

  /** Export user values to file. Returns backup metadata. */
  exportValues(userId: string, exportPath?: string): BackupMetadata {
    // Get user values
    const userValues = this._valueStore.getUserValues(userId);
    if (!userValues) {
      throw new Error(`User ${userId} not found`);
    }

    // Generate backup ID and path
    const backupId = crypto.randomUUID().slice(0, 8);

    const resolvedPath =
      exportPath ||
      path.join(
        this.backupDir,
        `${userId}_${new Date().toISOString().replace(/[:.]/g, "-").slice(0, 19)}.json`,
      );

    // Export data

    const exportData: Record<string, unknown> = {
      user_id: userId,
      exported_at: new Date().toISOString(),
      version: "2.2.0",
      values: userValueToDict(userValues),
    };

    // Write file
    const jsonStr = JSON.stringify(exportData, null, 2);
    fs.writeFileSync(resolvedPath, jsonStr, "utf-8");

    // Calculate simple checksum
    const checksum = String(Math.abs(hashCode(jsonStr))).slice(0, 16);

    // Create metadata
    const stat = fs.statSync(resolvedPath);
    const metadata: BackupMetadata = {
      userId,
      backupId,
      createdAt: new Date(),
      filePath: resolvedPath,
      fileSize: stat.size,
      valuesCount:
        userValues.principles.length +
        Object.keys(userValues.preferences).length +
        userValues.redLines.length,
      checksum,
    };

    // Save metadata
    this._saveMetadata(metadata);

    return metadata;
  }

  /** Import user values from file. Returns true if import was successful. */
  importValues(userId: string, importPath: string, overwrite: boolean = false): boolean {
    // Read file
    if (!fs.existsSync(importPath)) {
      throw new Error(`Backup file not found: ${importPath}`);
    }

    const content = fs.readFileSync(importPath, "utf-8");
    const data = JSON.parse(content) as Record<string, unknown>;

    if (!data.values) {
      throw new Error("Invalid backup file format");
    }

    const importedValues = data.values as Record<string, unknown>;

    // Get existing values
    const existing = this._valueStore.getUserValues(userId);

    if (existing && !overwrite) {
      throw new Error(`User ${userId} already exists. Use overwrite=true to replace.`);
    }

    // Import principles
    const principles = (importedValues.principles as string[]) || [];
    for (const principle of principles) {
      this._valueStore.savePrinciple(userId, principle);
    }

    // Import preferences
    const preferences = (importedValues.preferences as Record<string, unknown>) || {};
    for (const [key, value] of Object.entries(preferences)) {
      this._valueStore.savePreference(userId, key, value);
    }

    // Import red lines
    const redLines = (importedValues.red_lines as string[]) || [];
    for (const line of redLines) {
      this._valueStore.saveRedLine(userId, line);
    }

    return true;
  }

  /** List backup files. */
  listBackups(userId?: string): BackupMetadata[] {
    const metadataList: BackupMetadata[] = [];

    let allMetadata: Record<string, Record<string, unknown>> = {};
    if (fs.existsSync(this._metadataFile)) {
      try {
        allMetadata = JSON.parse(
          fs.readFileSync(this._metadataFile, "utf-8"),
        ) as Record<string, Record<string, unknown>>;
      } catch {
        allMetadata = {};
      }
    }

    for (const [backupId, metaData] of Object.entries(allMetadata)) {
      if (userId === undefined || metaData.user_id === userId) {
        try {
          metadataList.push(backupMetadataFromDict(metaData));
        } catch {
          // skip invalid entries
        }
      }
    }

    // Sort by time descending
    metadataList.sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime());

    return metadataList;
  }

  /** Get user backup metadata summary. */
  backupMetadata(userId: string): Record<string, unknown> {
    const backups = this.listBackups(userId);

    if (backups.length === 0) {
      return {
        user_id: userId,
        backup_count: 0,
        latest_backup: null,
        total_size: 0,
      };
    }

    return {
      user_id: userId,
      backup_count: backups.length,
      latest_backup: backupMetadataToDict(backups[0]),
      total_size: backups.reduce((sum, b) => sum + b.fileSize, 0),
    };
  }

  /** Delete backup by ID. Returns true if deletion was successful. */
  deleteBackup(backupId: string): boolean {
    let allMetadata: Record<string, Record<string, unknown>> = {};
    if (fs.existsSync(this._metadataFile)) {
      try {
        allMetadata = JSON.parse(
          fs.readFileSync(this._metadataFile, "utf-8"),
        ) as Record<string, Record<string, unknown>>;
      } catch {
        return false;
      }
    } else {
      return false;
    }

    if (!(backupId in allMetadata)) return false;

    const metaData = allMetadata[backupId];
    const filePath = metaData.file_path as string;

    // Delete file
    if (fs.existsSync(filePath)) {
      fs.unlinkSync(filePath);
    }

    // Delete metadata
    delete allMetadata[backupId];
    fs.writeFileSync(this._metadataFile, JSON.stringify(allMetadata, null, 2), "utf-8");

    return true;
  }

  /** Save backup metadata. */
  private _saveMetadata(metadata: BackupMetadata): void {
    let allMetadata: Record<string, Record<string, unknown>> = {};
    if (fs.existsSync(this._metadataFile)) {
      try {
        allMetadata = JSON.parse(
          fs.readFileSync(this._metadataFile, "utf-8"),
        ) as Record<string, Record<string, unknown>>;
      } catch {
        allMetadata = {};
      }
    }

    allMetadata[metadata.backupId] = backupMetadataToDict(metadata);
    fs.writeFileSync(this._metadataFile, JSON.stringify(allMetadata, null, 2), "utf-8");
  }
}

/** Simple string hash function for checksums. */
function hashCode(str: string): number {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const chr = str.charCodeAt(i);
    hash = (hash << 5) - hash + chr;
    hash |= 0; // Convert to 32bit integer
  }
  return hash;
}
